### PR TITLE
[spmd_types] switch full_dtensor to LP autograd fn

### PR DESCRIPTION
**Pull request by pianpwk**
Moves full_dtensor backend off pytorch's loss_parallel -> our custom autograd fn

Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.12.0) (oldest at bottom):
* #3673
* #3655
* #3654
* #3653
* #3278
* #3672
* #3472
* #3632
* __->__ #3671

### CODE REVIEW

File: tests/unit_tests/test_loss.py [hunk 1]
```diff
@@ -288,12 +288,12 @@ def test_loss_parallel_cross_entropy_parity(self):
         Tests loss-parallel cross-entropy loss bitwise parity, with torch.distributed.tensor.parallel.loss_parallel().
         Tests even/uneven vocab sharding, TP, DP+TP, and IGNORE_INDEX labels.
 
-        Runs _LossParallelCrossEntropy under typing checking: logits S(1)@TP, labels I@TP -> I@TP.
```
Comment (pianpwk) [inline tests/unit_tests/test_loss.py:289]: original test was using 2d logits, switch to [B, L, D] as loss.py is hardcoded for 3d logits

**Review by tianyu-l — APPROVED** [review]
Is this bitwise E2E with dtensor loss_parallel?

**Comment by pianpwk** [comment]
> Is this bitwise E2E with dtensor loss_parallel?

I was dumb, that only did it for non-chunked loss path. Now does it for chunked loss too, verified w/ llama3 & qwen3 x (TP, FSDP+TP, CP+TP), SP on/off, loss & grad norm 20 step
